torch/csrc/jit/tensorexpr/analysis.h

```diff
@@ -248,6 +248,110 @@ class ModifiesVarChecker : public IRVisitor {
   bool found_{false};
 };
 
+// Traverse the Block stmt to identify the live range of the specified buf. The
+// live range, indicated by a pair of integers, specifies the first and last
+// stmt in block stmts that access to the buf.
+class BufLiveRange : public IRVisitor {
+ public:
+  BufLiveRange(BufPtr b) : buf_(b) {}
+
+  std::tuple<int32_t, int32_t> getLiveRange() {
+    return std::make_tuple(begin_, end_);
+  }
+
+  static std::tuple<int32_t, int32_t> liveRange(StmtPtr s, BufPtr b) {
+    BlockPtr block = to<Block>(s);
+    TORCH_INTERNAL_ASSERT(
+        block, "Only analze buffer live range for block stmt");
+    BufLiveRange analyzer(b);
+    block->accept(&analyzer);
+    return analyzer.getLiveRange();
+  }
+
+ private:
+  bool findBufReads(StmtPtr s) {
+    auto loads1 = NodeFinder<Load>::find(s);
+    for (auto l : loads1) {
+      if (l->buf() == buf_) {
+        return true;
+      }
+    }
+    auto loads2 = NodeFinder<ExternalCall>::find(s);
+    for (auto l : loads2) {
+      for (auto lb : l->buf_args()) {
+        if (lb == buf_) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  bool findBufWrites(StmtPtr s) {
+    auto writes1 = NodeFinder<Store>::find(s);
+    for (auto w : writes1) {
+      if (w->buf() == buf_) {
+        return true;
+      }
+    }
+    auto writes2 = NodeFinder<ExternalCall>::find(s);
+    for (auto w : writes2) {
+      if (w->buf() == buf_) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void updateRange(StmtPtr s) {
+    bool has_reads = findBufReads(s), has_writes = findBufWrites(s);
+    if (has_reads || has_writes) {
+      if (begin_ == -1) {
+        begin_ = curr_index_;
+      };
+      end_ = curr_index_;
+    }
+  }
+
+  void visit(StorePtr v) {
+    updateRange(v);
+  }
+
+  void visit(LetPtr v) {
+    updateRange(v);
+  }
+
+  void visit(AtomicAddPtr v) {
+    updateRange(v);
+  }
+
+  void visit(ExternalCallPtr v) {
+    updateRange(v);
+  }
+
+  void visit(BlockPtr v) {
+    // Stmt counter of the block
+    bool enable_count = false;
+    for (StmtPtr s : *v) {
+      if (to_count_) {
+        curr_index_ += 1;
+        to_count_ = false;
+        enable_count = true;
+      }
+      s->accept(this);
```

Comment:
What's the point of using visitor here? Can we analyze `s` directly? (Essentially, can we call "updateRange" here instead of going through `::visit` calls? If we do this, we probably need to rename `updateRange`).

Also, I think this implementation might yield incorrect results if we have nested blocks (as visitor for inner blocks will rewrite `to_count_` and `cur_index_` values).

Reply:
`to_count_` cannot be rewritten by inner blocks because it's guarded by `enable_count`: `enable_count` is true only when we access the outermost loops. I have tests with nested blocks and they yield correct results.

Currently, `NodeFinder` is called (to update buf liveness range) when there are stmts like `store/external call`s. I agree that it would be cleaner to call `NodeFinder` for outer loops instead so we don't need to go through the stmts by ourselves. Will make the changes.

Reply:
For `updateRange`: it is short for `findBufAccessesAndUpdateBufLiveRange`. Do you have other name suggestions?

Reply:
> For updateRange: it is short for findBufAccessesAndUpdateBufLiveRange. Do you have other name suggestions?

I personally would probably go with something generic in this case, like `processStmt`. Note that "range" is already implied there because it's a method of `BufLiveRange` class, so there is not much value in repeating it in the method name (and it even can be confusing as it could mean that there are several different ranges we work with).

Reply:
> I agree that it would be cleaner to call NodeFinder for outer loops instead so we don't need to go through the stmts by ourselves. Will make the changes.

Ah, I see how it works. Yes, I think it can become much simpler and easier to follow after your changes!